### PR TITLE
Refactoring clustering/packing code and building re-cluster API

### PR DESCRIPTION
**Pull request by MohamedElgammal**
#### Description

In this pull request, the first function of re-cluster is implemented. This API should allow the users to change clustering decisions after the clustering is done. These changes can be done during packing or placement.

Multiple refactoring steps were done to the original clustering code to build this API. 

The first function in the API implemented in this PR can move an atom out of a cluster and create a new cluster for it. More Functions for this API are to be implemented in upcoming PRs.

The function was tested on VTR benchmarks and unit tests will be added when the API is completed in the upcoming PRs.

#### Related Issue

Being not able to change the clustering decision later in the flow caused suboptimality in the generated placed design.

#### How Has This Been Tested?

The refactor shows no change to QoR. Some examples are extracted for the CI runs comparing the branch to master with no change attached [here](https://docs.google.com/spreadsheets/d/1FUboB9A2ssrK9nM_6FU2iIkN3Fi3eAsGuVbx-hCk7_M/edit?usp=sharing)
#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [ x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x ] All new and existing tests passed

### CODE REVIEW

**Comment by MohamedElgammal** [comment]
All the current check failures are due to cloud resource issues. I will re-run the failing tests in the evening (tried that yesterday and all the tests passed). It seems that the cloud is less loaded in the evening. If it's green, I will merge it in the early morning tomorrow.

File: vpr/src/pack/pack.cpp [hunk 1]
```diff
@@ -41,31 +42,34 @@ bool try_pack(t_packer_opts* packer_opts,
...
               std::vector<t_lb_type_rr_node>* lb_type_rr_graphs) {
+    auto& helper_ctx = g_vpr_ctx.mutable_helper();
+
     std::unordered_set<AtomNetId> is_clock;
-    std::multimap<AtomBlockId, t_pack_molecule*> atom_molecules;                     //The molecules associated with each atom block
     std::unordered_map<AtomBlockId, t_pb_graph_node*> expected_lowest_cost_pb_gnode; //The molecules associated with each atom block
     const t_model* cur_model;
-    int num_models;
+    t_clustering_data clustering_data;
+    //int num_models;
```
Comment (sfkhalid) [inline vpr/src/pack/pack.cpp:51]: maybe remove these lines instead of commenting them

File: vpr/src/pack/prepack.cpp [hunk 2]
```diff
@@ -938,11 +927,13 @@ static void free_pack_pattern_block(t_pack_pattern_block* pattern_block, t_pack_
  * Side Effect: If successful, link atom to molecule
  */
 static t_pack_molecule* try_create_molecule(t_pack_patterns* list_of_pack_patterns,
-                                            std::multimap<AtomBlockId, t_pack_molecule*>& atom_molecules,
                                             const int pack_pattern_index,
                                             AtomBlockId blk_id) {
     t_pack_molecule* molecule;
 
+    //auto& atom_ctx = g_vpr_ctx.atom();
```
Comment (sfkhalid) [inline vpr/src/pack/prepack.cpp:934]: remove commented line

File: vpr/src/pack/prepack.h [hunk 3]
```diff
@@ -16,10 +16,9 @@ void free_list_of_pack_patterns(std::vector<t_pack_patterns>& list_of_pack_patte
 void free_pack_pattern(t_pack_patterns* pack_pattern);
 
 t_pack_molecule* alloc_and_load_pack_molecules(t_pack_patterns* list_of_pack_patterns,
-                                               std::multimap<AtomBlockId, t_pack_molecule*>& atom_molecules,
                                                std::unordered_map<AtomBlockId, t_pb_graph_node*>& expected_lowest_cost_pb_gnode,
                                                const int num_packing_patterns);
 
-void free_pack_molecules(t_pack_molecule* list_of_pack_molecules);
+//void free_pack_molecules(t_pack_molecule* list_of_pack_molecules);
```
Comment (sfkhalid) [inline vpr/src/pack/prepack.h:22]: remove commented line

File: vpr/src/pack/re_cluster_util.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+#include "clustered_netlist_fwd.h"
+#include "clustered_netlist_utils.h"
+#include "atom_netlist_fwd.h"
+#include "globals.h"
+#include "pack_types.h"
+#include "cluster_util.h"
+/**
+ * @file
+ * @brief This files defines some helper functions for the re-clustering
+ *
```
Comment (sfkhalid) [inline vpr/src/pack/re_cluster_util.h:13]: Maybe add some context to this comment - why would we want to re-do clusters, what situations would that be beneficial in?

File: vpr/src/place/place.cpp [hunk 5]
```diff
@@ -1057,6 +1063,14 @@ static void placement_inner_loop(const t_annealing_state* state,
                                              blocks_affected, delay_model, criticalities, setup_slacks,
                                              placer_opts, move_type_stat, place_algorithm, timing_bb_factor, manual_move_enabled);
 
+        /*
```
Comment (sfkhalid) [inline vpr/src/place/place.cpp:1066]: Remove this debugging code

**Review by sfkhalid — APPROVED** [review]
The code looks good to me overall. I think you can remove some unnecessary lines in a new pull request, and also add more context to the comment about the re-clustering API.

**Comment by MohamedElgammal** [comment]
Thanks @sfkhalid  for the review. As per our conversation yesterday, I will land this now and address your comments on a new PR so that you can work on debugging the tests on the master branch.

File: vpr/src/pack/re_cluster_util.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                              std::vector<t_lb_type_rr_node>* lb_type_rr_graphs,
+                              ClusterBlockId& old_clb,
+                              t_clustering_data& clustering_data,
+                              int& imacro,
+                              bool during_packing);
+
+/**
+ * @brief A function that starts a new cluster for one specific molecule
+ * 
+ * It place the molecule in a specific type and mode that should be passed by
```
Comment (vaughnbetz) [inline vpr/src/pack/re_cluster_util.h:55]: Same comment : comment inputs, outputs, etc.
inputs first, outputs second in parameter list.
Do we need pad_loc_type?
You're passing in an atom, but the comment is about molecules. (update comment if necessary)

File: vpr/src/pack/re_cluster_util.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                                bool enable_pin_feasibility_filter,
+                                ClusterBlockId clb_index,
+                                t_lb_router_data** router_data,
+                                std::vector<t_lb_type_rr_node>* lb_type_rr_graphs,
+                                PartitionRegion& temp_cluster_pr,
+                                t_clustering_data& clustering_data,
+                                bool during_packing);
+
+/**
+ * @brief A function that fix the clustered netlist if the move is performed
```
Comment (vaughnbetz) [inline vpr/src/pack/re_cluster_util.h:73]: If you can, explain what this fix is (at a high level).
When should this be called?  After any change to any cluster after packing is done?

**Review by vaughnbetz — COMMENTED** [review]
To apply to the latest code.
Sarah's unit tests: seem to be full flow tests --> maybe better as command line reg tests as usual?

**Pull request merged**